hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java

```diff
@@ -494,20 +459,87 @@ private void handleRetry(Object cause) {
         operationService.invocationRegistry.deregister(this);
 
         if (future.interrupted) {
-            future.complete(INTERRUPTED_RESPONSE);
-        } else if (!future.complete(WAIT_RESPONSE)) {
-            logger.finest("Cannot retry " + toString() + ", because a different response is already set: " + future.response);
+            future.complete(INTERRUPTED);
         } else {
+            // fast retry for the first few invocations
             if (invokeCount < MAX_FAST_INVOCATION_COUNT) {
-                // fast retry for the first few invocations
                 operationService.asyncExecutor.execute(this);
             } else {
-                nodeEngine.getExecutionService().schedule(ASYNC_EXECUTOR, this, tryPauseMillis, MILLISECONDS);
+                if (invokeCount < MAX_FAST_INVOCATION_COUNT) {
+                    // fast retry for the first few invocations
+                    operationService.asyncExecutor.execute(this);
+                } else {
+                    nodeEngine.getExecutionService().schedule(ASYNC_EXECUTOR, this, tryPauseMillis, MILLISECONDS);
+                }
             }
         }
     }
 
-    boolean checkBackupTimeout(long timeoutMillis) {
+    /**
+     * Checks if this Invocation has received a heart-beat in time.
+     *
+     * If the response is already set, or if a heart-beat has been received in time, then the false is returned.
+     *
+     * If no heart-beat has been received, then the future.set is called with HEARTBEAT_TIMEOUT
+     * and true is returned.
+     *
+     * gets called from the monitor-thread
+     *
+     * @return true if there is a timeout detected, false otherwise.
+     */
+    boolean detectAndHandleTimeout(long heartbeatTimeoutMillis) {
+        HeartbeatTimeout heartbeatTimeout = detectTimeout(heartbeatTimeoutMillis);
+
+        if (heartbeatTimeout == YES) {
+            future.complete(HEARTBEAT_TIMEOUT);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    HeartbeatTimeout detectTimeout(long heartbeatTimeoutMillis) {
+        if (pendingResponse != VOID) {
+            // if there is a response, then we won't timeout.
+            return NO__RESPONSE_AVAILABLE;
```

Comment:
This is confusing, even though there is a response (comment says that), method returns `NO__RESPONSE_AVAILABLE`, this lowers code readability and also can cause erroneous cases for the future maintainers due to the naming.